Engine registrations

dgeb opened the pull request:
Splits registrations between Application and Engine classes. Registrations that have proven to be necessary for all engines, based on work in the [ember-engines](https://github.com/dgeb/ember-engines) addon, are moved from the Application class to the Engine class.

This eliminates the need for overrides in ember-engines, while tests still ensure that applications are initialized with the same entries in their registries as always. This does make the registry verification tests rather verbose, but I wanted to make sure nothing was dropped during this transition.

There is still work TBD re: synchronization between engine instances and their parents which could be handled either in this same PR or separately.

/cc @rwjblue

dgeb committed:
Introduce test helpers for verifying registry contents.

dgeb committed:
Verify registry built for Ember.Application.

dgeb committed:
Split registrations between Application and Engine classes.

Registrations that have proven to be necessary for all engines,
based on work in the ember-engines addon, are moved from the
Application class to the Engine class.

This eliminates the need for overrides in ember-engines, while tests
still ensure that applications are initialized with the same entries in
their registries as always.

rwjblue commented:
@dgeb - This looks good to me.  

> There is still work TBD re: synchronization between engine instances and their parents which could be handled either in this same PR or separately.

I'm in favor of landing things in smaller pull requests (like this one) if possible.  Can we button this one up, and land it then submit subsequent PRs?

homu commented:
:umbrella: The latest upstream changes (presumably 3639774) made this pull request unmergeable. Please resolve the merge conflicts.

dgeb commented:
> I'm in favor of landing things in smaller pull requests (like this one) if possible. Can we button this one up, and land it then submit subsequent PRs?

@rwjblue Sounds good. I've rebased and removed the "WIP" tag.

rwjblue commented:
Thanks @dgeb!
